### PR TITLE
Fix memory leak in backend_impl.cc caused by PyObject_GetAttr

### PR DESCRIPTION
**Pull request by JanuszL**
- not increasing reference count to the result of PyObject_GetAttr
  causes leaking of the python object. This PR moves to the usage of
  `getattr` to handle this automatically

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

#### Additional information
- Affected modules and functionalities:
  - backend_impl.cc

- Key points relevant for the review:
  - NA

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: dali/python/backend_impl.cc [hunk 1]
```diff
@@ -56,15 +56,11 @@ void __backend_impl_force_tls_align_fun(void) {}
 using namespace pybind11::literals; // NOLINT
 
 static void* ctypes_void_ptr(const py::object& object) {
-  PyObject *p_ptr = object.ptr();
-  if (!PyObject_HasAttr(p_ptr, PyUnicode_FromString("value"))) {
+  auto ptr_as_int = getattr(object, "value", py::cast<py::none>(Py_None));
+  if (ptr_as_int.is(py::cast<py::none>(Py_None))) {
```
Comment (banasraf) [inline dali/python/backend_impl.cc:60]: https://pybind11.readthedocs.io/en/stable/reference.html#_CPPv4NK10object_api7is_noneEv
`is_none()`?

File: dali/python/backend_impl.cc [hunk 2]
```diff
@@ -199,15 +195,11 @@ void FillTensorFromDlPack(py::capsule capsule, SourceDataType<SrcBackend> *batch
...
                              string layout) {
-  PyObject *p_ptr = object.ptr();
-  if (!PyObject_HasAttr(p_ptr, PyUnicode_FromString("__cuda_array_interface__"))) {
-    DALI_FAIL("Provided object doesn't support cuda array interface protocol")
-  }
-  PyObject *ptr_intf = PyObject_GetAttr(p_ptr, PyUnicode_FromString("__cuda_array_interface__"));
-  if (ptr_intf == Py_None) {
+  py::dict cu_a_interface = getattr(object, "__cuda_array_interface__",
+                                    py::cast<py::none>(Py_None));
+  if (cu_a_interface.is(py::cast<py::none>(Py_None))) {
```
Comment (banasraf) [inline dali/python/backend_impl.cc:200]: https://pybind11.readthedocs.io/en/stable/reference.html#_CPPv4NK10object_api7is_noneEv
`is_none()`?

File: dali/python/backend_impl.cc [hunk 3]
```diff
@@ -56,15 +56,11 @@ void __backend_impl_force_tls_align_fun(void) {}
 using namespace pybind11::literals; // NOLINT
 
 static void* ctypes_void_ptr(const py::object& object) {
-  PyObject *p_ptr = object.ptr();
-  if (!PyObject_HasAttr(p_ptr, PyUnicode_FromString("value"))) {
+  auto ptr_as_int = getattr(object, "value", py::cast<py::none>(Py_None));
```
Comment (banasraf) [inline dali/python/backend_impl.cc:59]: `py::none()` should be equivalent with `py::cast<py::none>(Py_None)`

File: dali/python/backend_impl.cc [hunk 4]
```diff
@@ -199,15 +195,11 @@ void FillTensorFromDlPack(py::capsule capsule, SourceDataType<SrcBackend> *batch
...
 void FillTensorFromCudaArray(const py::object object, TensorType *batch, int device_id,
                              string layout) {
-  PyObject *p_ptr = object.ptr();
-  if (!PyObject_HasAttr(p_ptr, PyUnicode_FromString("__cuda_array_interface__"))) {
-    DALI_FAIL("Provided object doesn't support cuda array interface protocol")
-  }
-  PyObject *ptr_intf = PyObject_GetAttr(p_ptr, PyUnicode_FromString("__cuda_array_interface__"));
-  if (ptr_intf == Py_None) {
+  py::dict cu_a_interface = getattr(object, "__cuda_array_interface__",
+                                    py::cast<py::none>(Py_None));
```
Comment (banasraf) [inline dali/python/backend_impl.cc:199]: `py::none()` should be equivalent with `py::cast<py::none>(Py_None)`

File: dali/python/backend_impl.cc (hunk 1 above)
Comment (JanuszL) [reply dali/python/backend_impl.cc:60]: Done

File: dali/python/backend_impl.cc (hunk 3 above)
Comment (JanuszL) [reply dali/python/backend_impl.cc]: Done

File: dali/python/backend_impl.cc (hunk 4 above)
Comment (JanuszL) [reply dali/python/backend_impl.cc]: Done

**Comment by JanuszL** [comment]
!build

File: dali/python/backend_impl.cc (hunk 2 above)
Comment (JanuszL) [reply dali/python/backend_impl.cc]: Done

**Review by banasraf — APPROVED** [review]

**Review by jantonguirao — APPROVED** [review]

**Pull request merged**